lib/Support/PrettyPrinter.cpp

```diff
@@ -0,0 +1,299 @@
+        if (scanStack.empty())
+          return print(f);
+        tokens.push_back(f);
+        rightTotal += f.size;
+        assert(rightTotal > 0);
+        checkStream();
+      })
+      .Case([&](BreakToken *b) {
+        if (scanStack.empty())
+          clear();
+        else
+          checkStack();
+        addScanToken(-rightTotal);
+        rightTotal += b->spaces();
+        assert(rightTotal > 0);
+      })
+      .Case([&](BeginToken *b) {
+        if (scanStack.empty())
+          clear();
+        addScanToken(-rightTotal);
+      })
+      .Case([&](EndToken *end) {
+        if (scanStack.empty())
+          return print({t, 0});
+        addScanToken(-1);
+      });
+  rebaseIfNeeded();
+}
+
+void PrettyPrinter::rebaseIfNeeded() {
+  // Check for too-large totals, reset.
+  // This can happen if we have an open group and emit
+  // many tokens, especially newlines which have artificial size.
+  if (tokens.empty())
+    return;
+  assert(leftTotal >= 0);
+  assert(rightTotal >= 0);
+  if (uint32_t(leftTotal) > rebaseThreshold) {
+    // Plan: reset leftTotal to '1', adjust all accordingly.
+    auto adjust = leftTotal - 1;
+    for (auto &scanIndex : scanStack) {
+      assert(scanIndex >= tokenOffset);
+      auto &t = tokens[scanIndex - tokenOffset];
+      if (isa<BreakToken, BeginToken>(&t.token)) {
+        if (t.size < 0) {
+          assert(t.size + adjust < 0);
+          t.size += adjust;
+        }
+      }
+      // While walking, reset tokenOffset too.
+      scanIndex -= tokenOffset;
+    }
+    leftTotal -= adjust;
+    rightTotal -= adjust;
+    tokenOffset = 0;
+  }
+}
+
+void PrettyPrinter::eof() {
+  if (!scanStack.empty()) {
+    checkStack();
+    advanceLeft();
+  }
+  assert(scanStack.empty() && "unclosed groups at EOF");
+  if (scanStack.empty())
+    clear();
+}
+
+void PrettyPrinter::clear() {
+  assert(scanStack.empty() && "clearing tokens while still on scan stack");
+  assert(tokens.empty());
+  leftTotal = rightTotal = 1;
+  tokens.clear();
+  tokenOffset = 0;
+  if (listener)
+    listener->clear();
+}
+
+/// Break encountered, set sizes of begin/breaks in scanStack that we now know.
+void PrettyPrinter::checkStack() {
+  unsigned depth = 0;
+  while (!scanStack.empty()) {
+    auto x = scanStack.back();
+    assert(x >= tokenOffset && tokens.size() + tokenOffset > x);
+    auto &t = tokens[x - tokenOffset];
+    if (auto *b = llvm::dyn_cast<BeginToken>(&t.token)) {
+      if (depth == 0)
+        break;
+      scanStack.pop_back();
+      t.size += rightTotal;
+      --depth;
+    } else if (auto *e = llvm::dyn_cast<EndToken>(&t.token)) {
+      scanStack.pop_back();
+      t.size = 1;
+      ++depth;
+    } else {
+      scanStack.pop_back();
+      t.size += rightTotal;
+      if (depth == 0)
+        break;
+    }
+  }
+}
+
+/// Check if there's enough tokens to hit width, if so print.
+/// If scan size is wider than line, it's infinity.
```

Comment:
Mega nit: "there're"

Reply:
Not sure I follow, which should be "there're"?  "there's" is meant to be "there is" and "it's" is describing that the size is "infinity" if we already have seen enough to fill this line--don't actually need to wait until have all tokens for size computation once it's over line length.

LMK if this could be clearer, thanks!

Reply:
Mega grammar nit (I think!): `Check if there are tokens` vs. `Check if there is tokens`. This could use the contraction `there're`, though it's somewhat non-standard and difficult to say/read.

Reply:
Oh, wow, yes.  Haha, thanks for ..spelling it out! :wink:.

Reply:
(Fixed, thanks!)